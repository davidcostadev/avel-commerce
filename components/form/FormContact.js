import React from 'react'
import PropTypes from 'prop-types'
import FormGroup from '../form/FormGroup'
import Input from '../form/Input'
import Textarea from '../form/Textarea'
import { FlashMessages } from '../page/FlashMessage'

class FormContact extends React.Component {
  constructor(props) {
    super(props)

    this.onSubmit = this.onSubmit.bind(this)
    this.handle = this.handle.bind(this)

    this.state = {
      name: '',
      email: '',
      subject: '',
      message: '',
      isLoading: false,
      errors: [],
    }
  }

  onSubmit(event) {
    event.preventDefault()

    const {
      name,
      email,
      subject,
      message,
    } = this.state

    if (!name.length) {
      this.addErrors({
        id: 1,
        type: 'danger',
        message: 'Por favor, informe o nome',
      })
    } else if (!email.length) {
      this.addErrors({
        id: 2,
        type: 'danger',
        message: 'Por favor, informe o email',
      })
    } else if (!subject.length) {
      this.addErrors({
        id: 3,
        type: 'danger',
        message: 'Por favor, informe o assunto',
      })
    } else if (!message.length) {
      this.addErrors({
        id: 4,
        type: 'danger',
        message: 'Por favor, informe sua mensagem',
      })
    } else {
      this.setState({
        errors: [],
      })
      this.props.onChange({
        name,
        email,
        subject,
        message,
      })
    }
  }

  handle({ target }) {
    this.setState({
      [target.id]: target.value,
    })
  }

  addErrors({ id, type, message }) {
    this.setState({
      errors: [
        {
          id,
          type,
          message,
        },
      ],
    })
  }

  render() {
    const { isLoading, errors } = this.state

    return (
      <form>
        <FlashMessages msgs={errors} />
        <FormGroup>
          <Input
            name="name"
            id="name"
            placeholder="Seu Nome"
            onChange={this.handle}
          />
        </FormGroup>
        <FormGroup>
          <Input
            type="email"
            name="email"
            id="email"
            placeholder="Digite seu email"
            onChange={this.handle}
          />
        </FormGroup>
        <FormGroup>
          <Input
            name="subject"
            id="subject"
            placeholder="Assunto"
            onChange={this.handle}
          />
        </FormGroup>
        <FormGroup>
          <Textarea
            name="message"
            id="message"
            placeholder="Mensagem"
            onChange={this.handle}
          />
        </FormGroup>
        <div className="text-center">
          <button
            className="btn btn-primary"
            disabled={isLoading}
            onClick={this.onSubmit}
          >
            Enviar
          </button>
        </div>
      </form>
    )
  }
}

FormContact.propTypes = {
  onChange: PropTypes.func.isRequired,
}

export default FormContact
