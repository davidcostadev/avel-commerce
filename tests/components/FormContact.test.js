/* eslint no-undef: "off" */
import React from 'react'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import renderer from 'react-test-renderer'
import FormContact from '../../components/form/FormContact'

Enzyme.configure({ adapter: new Adapter() })

describe('should', () => {
  it('Add product on cart', () => {
    const onChange = jest.fn()

    const tree = renderer.create((
      <FormContact
        onChange={onChange}
      />
    )).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('fill form and submit', () => {
    const onChange = jest.fn()
    const preventDefault = jest.fn()

    const wrapper = shallow((
      <FormContact
        onChange={onChange}
      />
    ))


    wrapper.find('button').simulate('click', { preventDefault })
    expect(wrapper.get(0)).toMatchSnapshot()

    wrapper.find('[id="name"]').simulate('change', {
      target: {
        id: 'name',
        value: 'david',
      },
    })

    wrapper.find('button').simulate('click', { preventDefault })
    expect(wrapper.get(0)).toMatchSnapshot()

    wrapper.find('[id="email"]').simulate('change', {
      target: {
        id: 'email',
        value: 'david@email.com',
      },
    })

    wrapper.find('button').simulate('click', { preventDefault })
    expect(wrapper.get(0)).toMatchSnapshot()

    wrapper.find('[id="subject"]').simulate('change', {
      target: {
        id: 'subject',
        value: 'subject',
      },
    })
    wrapper.find('button').simulate('click', { preventDefault })
    expect(wrapper.get(0)).toMatchSnapshot()

    wrapper.find('[id="message"]').simulate('change', {
      target: {
        id: 'message',
        value: 'message',
      },
    })

    wrapper.find('button').simulate('click', { preventDefault })
    expect(preventDefault).toBeCalled()
    expect(wrapper.get(0)).toMatchSnapshot()
    expect(onChange).toBeCalled()
    expect(onChange).toMatchSnapshot()
  })
})
