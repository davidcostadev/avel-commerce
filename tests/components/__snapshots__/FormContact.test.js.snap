// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should Add product on cart 1`] = `
<form>
  <div
    className="form-group"
  >
    <input
      className="form-control"
      id="name"
      name="name"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Seu Nome"
      type="text"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <input
      className="form-control"
      id="email"
      name="email"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Digite seu email"
      type="email"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <input
      className="form-control"
      id="subject"
      name="subject"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Assunto"
      type="text"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <textarea
      className="form-control"
      id="message"
      name="message"
      onChange={[Function]}
      placeholder="Mensagem"
      value=""
    />
  </div>
  <div
    className="text-center"
  >
    <button
      className="btn btn-primary"
      disabled={false}
      onClick={[Function]}
    >
      Enviar
    </button>
  </div>
</form>
`;

exports[`should fill form and submit 1`] = `
<form>
  <FlashMessages
    msgs={
      Array [
        Object {
          "id": 1,
          "message": "Por favor, informe o nome",
          "type": "danger",
        },
      ]
    }
  />
  <FormGroup
    row={false}
  >
    <Input
      id="name"
      name="name"
      onChange={[Function]}
      placeholder="Seu Nome"
      value=""
    />
  </FormGroup>
  <FormGroup
    row={false}
  >
    <Input
      id="email"
      name="email"
      onChange={[Function]}
      placeholder="Digite seu email"
      type="email"
      value=""
    />
  </FormGroup>
  <FormGroup
    row={false}
  >
    <Input
      id="subject"
      name="subject"
      onChange={[Function]}
      placeholder="Assunto"
      value=""
    />
  </FormGroup>
  <FormGroup
    row={false}
  >
    <Textarea
      id="message"
      name="message"
      onChange={[Function]}
      placeholder="Mensagem"
      value=""
    />
  </FormGroup>
  <div
    className="text-center"
  >
    <button
      className="btn btn-primary"
      disabled={false}
      onClick={[Function]}
    >
      Enviar
    </button>
  </div>
</form>
`;

exports[`should fill form and submit 2`] = `
<form>
  <FlashMessages
    msgs={
      Array [
        Object {
          "id": 2,
          "message": "Por favor, informe o email",
          "type": "danger",
        },
      ]
    }
  />
  <FormGroup
    row={false}
  >
    <Input
      id="name"
      name="name"
      onChange={[Function]}
      placeholder="Seu Nome"
      value=""
    />
  </FormGroup>
  <FormGroup
    row={false}
  >
    <Input
      id="email"
      name="email"
      onChange={[Function]}
      placeholder="Digite seu email"
      type="email"
      value=""
    />
  </FormGroup>
  <FormGroup
    row={false}
  >
    <Input
      id="subject"
      name="subject"
      onChange={[Function]}
      placeholder="Assunto"
      value=""
    />
  </FormGroup>
  <FormGroup
    row={false}
  >
    <Textarea
      id="message"
      name="message"
      onChange={[Function]}
      placeholder="Mensagem"
      value=""
    />
  </FormGroup>
  <div
    className="text-center"
  >
    <button
      className="btn btn-primary"
      disabled={false}
      onClick={[Function]}
    >
      Enviar
    </button>
  </div>
</form>
`;

exports[`should fill form and submit 3`] = `
<form>
  <FlashMessages
    msgs={
      Array [
        Object {
          "id": 3,
          "message": "Por favor, informe o assunto",
          "type": "danger",
        },
      ]
    }
  />
  <FormGroup
    row={false}
  >
    <Input
      id="name"
      name="name"
      onChange={[Function]}
      placeholder="Seu Nome"
      value=""
    />
  </FormGroup>
  <FormGroup
    row={false}
  >
    <Input
      id="email"
      name="email"
      onChange={[Function]}
      placeholder="Digite seu email"
      type="email"
      value=""
    />
  </FormGroup>
  <FormGroup
    row={false}
  >
    <Input
      id="subject"
      name="subject"
      onChange={[Function]}
      placeholder="Assunto"
      value=""
    />
  </FormGroup>
  <FormGroup
    row={false}
  >
    <Textarea
      id="message"
      name="message"
      onChange={[Function]}
      placeholder="Mensagem"
      value=""
    />
  </FormGroup>
  <div
    className="text-center"
  >
    <button
      className="btn btn-primary"
      disabled={false}
      onClick={[Function]}
    >
      Enviar
    </button>
  </div>
</form>
`;

exports[`should fill form and submit 4`] = `
<form>
  <FlashMessages
    msgs={
      Array [
        Object {
          "id": 4,
          "message": "Por favor, informe sua mensagem",
          "type": "danger",
        },
      ]
    }
  />
  <FormGroup
    row={false}
  >
    <Input
      id="name"
      name="name"
      onChange={[Function]}
      placeholder="Seu Nome"
      value=""
    />
  </FormGroup>
  <FormGroup
    row={false}
  >
    <Input
      id="email"
      name="email"
      onChange={[Function]}
      placeholder="Digite seu email"
      type="email"
      value=""
    />
  </FormGroup>
  <FormGroup
    row={false}
  >
    <Input
      id="subject"
      name="subject"
      onChange={[Function]}
      placeholder="Assunto"
      value=""
    />
  </FormGroup>
  <FormGroup
    row={false}
  >
    <Textarea
      id="message"
      name="message"
      onChange={[Function]}
      placeholder="Mensagem"
      value=""
    />
  </FormGroup>
  <div
    className="text-center"
  >
    <button
      className="btn btn-primary"
      disabled={false}
      onClick={[Function]}
    >
      Enviar
    </button>
  </div>
</form>
`;

exports[`should fill form and submit 5`] = `
<form>
  <FlashMessages
    msgs={Array []}
  />
  <FormGroup
    row={false}
  >
    <Input
      id="name"
      name="name"
      onChange={[Function]}
      placeholder="Seu Nome"
      value=""
    />
  </FormGroup>
  <FormGroup
    row={false}
  >
    <Input
      id="email"
      name="email"
      onChange={[Function]}
      placeholder="Digite seu email"
      type="email"
      value=""
    />
  </FormGroup>
  <FormGroup
    row={false}
  >
    <Input
      id="subject"
      name="subject"
      onChange={[Function]}
      placeholder="Assunto"
      value=""
    />
  </FormGroup>
  <FormGroup
    row={false}
  >
    <Textarea
      id="message"
      name="message"
      onChange={[Function]}
      placeholder="Mensagem"
      value=""
    />
  </FormGroup>
  <div
    className="text-center"
  >
    <button
      className="btn btn-primary"
      disabled={false}
      onClick={[Function]}
    >
      Enviar
    </button>
  </div>
</form>
`;

exports[`should fill form and submit 6`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "email": "david@email.com",
        "message": "message",
        "name": "david",
        "subject": "subject",
      },
    ],
  ],
}
`;
